lxd/storage/utils.go

```diff
@@ -360,3 +384,253 @@ func GetStorageResource(path string) (*api.ResourcesStoragePool, error) {
 
 	return &res, nil
 }
+
+// VolumeTypeNameToType converts a volume type string to internal code.
+func VolumeTypeNameToType(volumeTypeName string) (int, error) {
+	switch volumeTypeName {
+	case db.StoragePoolVolumeTypeNameContainer:
+		return db.StoragePoolVolumeTypeContainer, nil
+	case db.StoragePoolVolumeTypeNameImage:
+		return db.StoragePoolVolumeTypeImage, nil
+	case db.StoragePoolVolumeTypeNameCustom:
+		return db.StoragePoolVolumeTypeCustom, nil
+	}
+
+	return -1, fmt.Errorf("invalid storage volume type name")
+}
+
+// VolumeDBCreate creates a volume in the database.
+func VolumeDBCreate(s *state.State, poolName string, volumeName, volumeDescription string, volumeTypeName string, snapshot bool, volumeConfig map[string]string) error {
+	// Convert the volume type name to our internal integer representation.
+	volumeType, err := VolumeTypeNameToType(volumeTypeName)
+	if err != nil {
+		return err
+	}
+
+	// Load storage pool the volume will be attached to.
+	poolID, poolStruct, err := s.Cluster.StoragePoolGet(poolName)
+	if err != nil {
+		return err
+	}
+
+	// Check that a storage volume of the same storage volume type does not
+	// already exist.
+	volumeID, _ := s.Cluster.StoragePoolNodeVolumeGetTypeID(volumeName, volumeType, poolID)
+	if volumeID > 0 {
+		return fmt.Errorf("a storage volume of type %s does already exist", volumeTypeName)
+	}
+
+	// Make sure that we don't pass a nil to the next function.
+	if volumeConfig == nil {
+		volumeConfig = map[string]string{}
+	}
+
+	// Validate the requested storage volume configuration.
+	err = VolumeValidateConfig(poolName, volumeConfig, poolStruct)
+	if err != nil {
+		return err
+	}
+
+	err = VolumeFillDefault(poolName, volumeConfig, poolStruct)
+	if err != nil {
+		return err
+	}
+
+	// Create the database entry for the storage volume.
+	_, err = s.Cluster.StoragePoolVolumeCreate("default", volumeName, volumeDescription, volumeType, snapshot, poolID, volumeConfig)
+	if err != nil {
+		return fmt.Errorf("Error inserting %s of type %s into database: %s", poolName, volumeTypeName, err)
+	}
+
+	return nil
+}
+
+// SupportedPoolTypes the types of pools supported.
+var SupportedPoolTypes = []string{"btrfs", "ceph", "cephfs", "dir", "lvm", "zfs"}
+
+// StorageVolumeConfigKeys config validation for btrfs, ceph, cephfs, dir, lvm, zfs types.
+var StorageVolumeConfigKeys = map[string]func(value string) ([]string, error){
```

Comment:
We're going to want that split between generic config options and driver specific config options.

The drivers will be responsible for validating their own config and no mention of those config keys should be present outside of the driver.

The backend will be responsible for handling the generic config options (such as `security.shifted` and `security.unmapped`) and we'll filter those out from the data exposed to the drivers so we don't need them all be aware of config options that don't concern them.

Reply:
@stgraber ok, so are you thinking that there would be a non-exported function attached to the lxdBackend called `volumePropertiesTranslate` which would then validate generic parts of the config, and inside that function call the storage pool's driver function `VolumePropertiesTranslate` function passing it a modifier config with the generic fields removed for the driver to validate/remove and then the return from that would be merged back into the generic fields and returned to the caller?

Reply:
Yeah, in my head, we'd just have:
 - A map in the `storage` package for the generic keys
 - A private `validateVolume` options in `lxdBackend` which ignore additional keys and chains to the public function from the driver
 - A public `ValidateVolume` function in the Driver which takes a `StorageVolumesPost` with its config filtered so the keys in the map are filtered out.

Reply:
@stgraber the approach I've gone with is similar, but more like the device validation.

I've added a `validateVolumeCommonRules` and `validateVolume` function to the common driver. The former generates the common validation rules, and the latter is called from each driver's exported `ValidateVolume` function. It will take the `validateVolumeCommonRules` generated rules as a base and then merge any custom rules supplied by the driver.

It will then run through all the validation rules inside the driver (both common and driver specific). 

This achieves the same effect, but just defines the config in the `common` driver rather than in `lxdBackend`.

Reply:
@stgraber I'm not clear as to whether the `size` property is valid for `dir` volumes. The defaults seem to populate it with an empty string, and the validation function doesnt include dir type when validating size. But project quotas suggest it should have a size property. Can you clarify?

Reply:
@stgraber confirmed on IRC that the size property is valid for dir, have added support for that now, and made the "size" property a common config field for all storage drivers.